text/3519-arbitrary-self-types-v2.md

```diff
@@ -0,0 +1,403 @@
+
+This feature is fundamentally aimed at smart pointer types `P<T>` where it's not safe to create a reference `&T`. As noted above, that's most commonly because of semantic differences to pointers in other languages, but it might be because references have special meaning or behavior in some pure Rust domain. Either way, it's not OK to create a Rust reference `&T` or `&mut T`, yet we may want to allow methods to be called on some reference-like thing.
+
+For this reason, implementing `Deref::deref` is problematic for _nearly everyone who wants to use arbitrary self types_.
+
+If you're implementing a smart pointer `P<T>` yet you can't allow a reference `&T` to exist, any option for implementing `Deref::deref` has drawbacks:
+
+* Specify `Deref::Target=T` and panic in `Deref::deref`. Not good.
+* Specify `Deref::Target=*const T`. This works with the current arbitrary self types feature, but is only possible if your smart pointer type contains a `*const T` which you can reference - this isn't the case for (for instance) weak pointers or types containing `NonNull`.
+
+Therefore, the current Arbitrary Self Types v2 provides a separate `Receiver` trait.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When declaring a method, users can declare the type of the `self` receiver to be any type `T` where `T: Receiver<Target = Self>` or `Self`.
+
+The `Receiver` trait is simple and only requires to specify the `Target` type:
+
+```rust
+trait Receiver {
+    type Target: ?Sized;
+}
+```
+
+The `Receiver` trait is already implemented for many standard library types:
+- smart pointers in the standard library: `Rc<Self>`, `Arc<Self>`, `Box<Self>`, and `Pin<Ptr<Self>>` (and in fact, any type which implements `Deref`)
+- references: `&Self` and `&mut Self`
+- pointers: `*const Self` and `*mut Self`
+
+Shorthand exists for references, so that `self` with no ascription is of type `Self`, `&self` is of type `&Self` and `&mut self` is of type `&mut Self`.
+
+All of the following self types are valid:
+
+```rust
+impl Foo {
+    fn by_value(self /* self: Self */);
+    fn by_ref(&self /* self: &Self */);
+    fn by_ref_mut(&mut self /* self: &mut Self */);
+    fn by_ptr(self: *const Self);
+    fn by_mut_ptr(self: *mut Self);
+    fn by_box(self: Box<Self>);
+    fn by_rc(self: Rc<Self>);
+    fn by_custom_ptr(self: CustomPtr<Self>);
+}
+
+struct CustomPtr<T>(*const T);
+
+impl<T> Receiver for CustomPtr<T> {
+    type Target = T;
+}
+```
+
+## Recursive arbitrary receivers
+
+Receivers are recursive and therefore allowed to be nested. If type `T` implements `Receiver<Target=U>`, and type `U` implements `Receiver<Target=Self>`, `T` is a valid receiver (and so on outward).
+
+For example, this self type is valid:
+
+```rust
+impl MyType {
+     fn by_rc_to_box(self: Rc<Box<Self>>) { ... }
+}
+```
+
```

Comment:
It's worth commenting on how this might mean that *using* the `Receiver` trait is unreliable if you're not the compiler, since `Receiver<Target: Receiver<Target = Self>>` generally means that `Receiver<Target != Self>` (using pseudo-syntax).

I can't think of a circumstance where someone would *want* to bound their code by `Receiver`, but if they are, they're probably doing something wrong.

Reply:
A similar restriction applies to `Deref`, so at least it's nothing new

Reply:
Right, but still worth reiterating IMHO.

Reply:
> A similar restriction applies to `Deref`, so at least it's nothing new

Is this restriction for Deref documented somewhere?

Reply:
> I can't think of a circumstance where someone would want to bound their code by `Receiver`, but if they are, they're probably doing something wrong.

I think this is in fact sort of what I'm hinting at in one of the suggestions about [diagnostics, below](https://github.com/rust-lang/rfcs/pull/3519/files#diff-085d4b194a1af0c50cdb08e0ec395782aab67202ffc9d626ca6574c480bcbc42R257).

Perhaps we could explicitly issue a warning for anyone who bounds anything by `Receiver`? WDYT?

Reply:
`Deref` has the same issue yet `Deref` bounds sometimes make sense, e.g. on [`Pin::new`](https://doc.rust-lang.org/std/pin/struct.Pin.html#impl-Pin%3CP%3E).

Reply:
> > A similar restriction applies to `Deref`, so at least it's nothing new
> 
> Is this restriction for Deref documented somewhere?

As far as I'm aware, not explicitly. But it's pretty simple to understand; think of the implementation for `Box<T>`:

```rust
impl<T> Deref for Box<T> {
    type Target = T;
    // ...
}
```

By this definition, we know that the `Target` for `Box<Box<T>>` is `Box<T>`. But the compiler is able to unwrap this definition to just target `T`, since it can apply `Deref` iteratively. This means that if, for example, you had the bound `Deref<Target = T>`, you could not pass `Box<Box<T>>`, since the target for this is `Box<T>`, not `T`.

I don't believe it's possible to specify a bound that would iteratively unwrap the target like this without creating a helper trait, and even then, it's not immediately clear how you'd write such a trait. Maybe there's a crate that does it that I don't know about.

Reply:
Thanks for this discussion. I've tried to summarize it in the RFC in 544f1afa.

Reply:
I'm going to mark this as Resolved but please do reopen if the above commit doesn't reflect things here. Thanks again!